In [1]:
import numpy as np

$$ f(x) = -\cos x_1^2 - \cos x_2^2 -x_1 - x_2$$
$$ \nabla f(x) = \left(\begin{array}{1}
2x_1 \sin x_1^2 - 1\\
2x_2 \sin x_2^2 - 1
\end{array}\right)$$

In [12]:
x = np.array([0., 1.])
b1 = 0.9
b2 = 0.999
g = np.array([0.,0.])
m = np.array([0.,0.])
v = np.array([0.,0.])
mm = np.array([0.,0.])
vv = np.array([0.,0.])
alpha = 0.25
for i in range(3):
    epsilon = pow(10, -8)
    g = 2. * x * np.sin(x ** 2) - np.array([1., 1.])
    m = b1 * m + (1.-b1) * g
    v = b2 * v + (1.-b2) * g * g
    mm = m/(1. - b1)
    vv = v/(1. - b2)
    x[0] = x[0] - alpha/(np.sqrt(vv[0]) + epsilon) * mm[0]
    x[1] = x[1] - alpha/(np.sqrt(vv[1]) + epsilon) * mm[1]
    print(i + 1, ':', x)

1 : [0.25 0.75]
2 : [0.58563967 0.60428229]
3 : [0.96256447 0.65791692]


$$x_1 = \left(\begin{array}{1}
0.25\\
0.75
\end{array}\right), x_2 = \left(\begin{array}{1}
0.59\\
0.60
\end{array}\right), x_3 = \left(\begin{array}{1}
0.96\\
0.66
\end{array}\right)$$